compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -643,6 +644,91 @@ struct DistributeTranspose final : OpDistributionPattern<vector::TransposeOp> {
   }
 };
 
+struct DistributeLayoutConflictResolutions final
+    : OpDistributionPattern<IREE::VectorExt::LayoutConflictResolutionOp> {
+  using OpDistributionPattern::OpDistributionPattern;
+
+  VectorValue reshapeVector(Location loc, RewriterBase &rewriter,
```

Comment:
I am not 100% sure how we are doing this reshaping. Let's say we have a layout:

<BATCHX: 1, LANEY: 2>, <BATCHY: 1, LANEY: 2, VECTORX: 4>

```
0 0 0 0 1 1 1 1
2 2 2 2 3 3 3 3
```

Converting it to <BATCHX: 2, LANEY: 2>, <BATCHY: 1, LANEX: 2, VECTORX: 2>:

```
0 0 1 1 0 0 1 1
2 2 3 3 2 2 3 3
```

Are we doing this? If so, it looks wrong. The mapping is not the same.

Reply:
Okay so couple of follow-ups after our earlier discussion - this is more general of a pattern than it seems. Essentially it handles the "permutation-invariant" layout resolution scenarios. Let me try and explain. So say we are changing the column layout. Since the lane dimension is the same for both layouts(2), that implies that the we are just permuting the data in a row. From your example above,
```
0 0 0 0 1 1 1 1
```
becomes
```
0 0 1 1 0 0 1 1
```
For all unary operations (such as exp2) or even reductions along a single axis (like sum, max), since the operations are permutation invariant you will end up with the correct answer. So even if thread 0 owns the data corresponding to `0011` and thread 1 owns the data corresponding to `0011` from the original layout, when you do the unary operation you will get the correct answer. Which means that for permutation invariant operators, when the lane dimension is the same, we can treat layout conflict resolutions as simple reshapes.

Now for the more interesting binary operations, you can see that this resolution will work for linalg generics with parallel iterator types as long as both operands are permuted exactly the same way. This can also work for inner and outer product matmuls as long as both operands are permuted the same way.

To see this, say originally, we had A and B.T
```
0 0 0 0 1 1 1 1
2 2 2 2 3 3 3 3
4 4 4 4 5 5 5 5
6 6 6 6 7 7 7 7
```

So if we permute A and B's columns we end up with
```
0 0 1 1 0 0 1 1
2 2 3 3 2 2 3 3
4 4 5 5 4 4 5 5
6 6 7 7 6 6 7 7
```
So even if we did an inner or outer product, since the multiplication element order is preserved you get the correct result. The final piece of the puzzle is that even though the permutations might make sure the correct elements are being reduced for us to get the final correct result, we need to make sure that our final layout is consistent with the layout of the output tensor. Anyway, my proposal is to keep this here but add comments about the permutation invariant nature of the resolution. What do you think?

Reply:
I added a comment explaining this.